modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentPerVersionStore.java

```diff
@@ -322,13 +322,33 @@ else if (log.isDebugEnabled())
         while (true) {
             List<SharedDeployment> depsToCheck = null;
 
-            SharedDeployment dep = null;
+            SharedDeployment dep;
 
             synchronized (mux) {
                 // Check obsolete request.
                 if (isDeadClassLoader(meta))
                     return null;
 
+                Collection<GridDeployment> created = getDeployments();
+
+                boolean skipSearchDeployment = false;
+
+                // Check already exist deployment.
+                if (meta.deploymentMode() == SHARED) {
+                    for (GridDeployment dep0 : created) {
+                        // hot redeploy from same node
+                        if (dep0.participants().containsKey(meta.senderNodeId()) || dep0.undeployed())
+                            continue;
+
+                        IgniteBiTuple<Class<?>, Throwable> cls = dep0.deployedClass(meta.className(), meta.alias());
+
+                        if (cls.getKey() != null && cls.getValue() == null) {
+                            addParticipant((SharedDeployment)dep0, meta);
+                            skipSearchDeployment = true;
+                        }
```

Comment:
Why don't bracke the loop when you find the deployment?

Reply:
Also, I think we can assigned it into dep variable and brake the root while loop.

Reply:
because in common - getDeployments() returns the collection an we need to update all participants